IsProductSalable conditions pool

Stepa4man opened the pull request:
Creating pool of configuration conditions in IsProductSalable Service

Stepa4man committed:
is product salable configurations

Stepa4man committed:
ststic test fix

maghamed commented on app/code/Magento/InventoryCatalog/Model/GetLegacyStockItem.php:
```diff
@@ -0,0 +1,76 @@
+
+/**
+ * Service GetLegacyStockItem returns old inventory item data
+ * @package Magento\InventoryCatalog\Model
```
please remove package doc block

Stepa4man replied:
Done

maghamed commented on app/code/Magento/InventoryCatalog/Model/GetLegacyStockItem.php:
```diff
@@ -0,0 +1,76 @@
+     */
+    public function execute(string $sku)
+    {
+        if (!$this->item) {
```
that's exactly the reason why services should be stateless. 
as your service accept $sku
but cache `$this->item`

So, these calls 
```
$sku1 = execute('SKU-1');
$sku2 = execute('SKU-2');
$sku1 === $sku2  // the same object
```

Please, let's remove this caching

Stepa4man replied:
Done

maghamed commented on app/code/Magento/InventoryCatalog/Model/IsBackOrderAvailable.php:
```diff
@@ -0,0 +1,49 @@
+
+/**
+ * Class IsBackOrderAvailable
+ * @package Magento\InventoryCatalog\Model
```
please remove package

Stepa4man replied:
Done

maghamed commented on app/code/Magento/InventoryConfiguration/Model/ConfigMinQty.php:
```diff
@@ -0,0 +1,81 @@
+
+/**
+ * Class ConfigMinQty
+ * @package Magento\InventoryConfiguration\Model
```
please remove package

Stepa4man replied:
Done

maghamed commented on app/code/Magento/InventoryConfiguration/Model/StockItemConditionInterface.php:
```diff
@@ -0,0 +1,22 @@
+
+/**
+ * Interface StockConditionInterface
+ * @package Magento\InventoryConfiguration\Model
```
please remove package

Stepa4man replied:
Done

maghamed commented on app/code/Magento/InventoryConfiguration/Model/StockItemConditionInterface.php:
```diff
@@ -0,0 +1,22 @@
+ * Interface StockConditionInterface
+ * @package Magento\InventoryConfiguration\Model
+ */
+interface StockItemConditionInterface
```
So... 
what you've implemented is additional service which is passed to isProductSalable instance. 

And what isProductSalable does is call all the Condition services 
```
 foreach ($this->conditions as $condition) {
        if (null === $stockItemData) {	            
                 if ($condition->match($sku, $stockId) === true) {
                return false;	
                return true;
       }	
```
isProductSalable looks and behaves like `Composite` object. So, the quetion why do we need to have 2 composites: isProductSalable and StockItemCondition? 

What we need to have - the only composite with a clean and understandable interface. Moreover we already have this interface. Its name is `isProductSalableInterface`

Thus, what we need to do - is remove `StockItemConditionInterface` and make all its implementations to implement `isProductSalableInterface` interface.
We will end up having 1 composite object isProductSalable which will accept an array of isProductSalableInterface[] and do similar operation 
```
 foreach ($this->services as $isProductSalable) {    
                 if ($isProductSalable->execute($sku, $stockId) === true) {
                     return true;	
      //...
       }	
```

Stepa4man replied:
Done

maghamed commented on app/code/Magento/InventoryCatalog/Model/IsBackOrderAvailable.php:
```diff
@@ -0,0 +1,49 @@
+    public function match(string $sku, int $stockItem): bool
+    {
+        $legacyStockItem = $this->getLegacyStockItem->execute($sku);
+        if ($legacyStockItem->getBackorders() != Stock::BACKORDERS_NO) {
```
please always use strict comparison,
so !== instead of != and  === instead of ==

Stepa4man replied:
Done

maghamed commented on app/code/Magento/InventoryCatalog/Model/GetLegacyStockItem.php:
```diff
@@ -0,0 +1,76 @@
+ * Service GetLegacyStockItem returns old inventory item data
+ * @package Magento\InventoryCatalog\Model
+ */
+class GetLegacyStockItem
```
we no need this service, we need new API for StockItemConfiguration.
Which should be located at InventoryConfigurationApi module

Stepa4man replied:
Done

maghamed commented on app/code/Magento/InventoryConfiguration/Model/ConfigMinQty.php:
```diff
@@ -0,0 +1,81 @@
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryConfiguration\Model;
```
should be moved to InventorySales

Stepa4man replied:
Done

maghamed commented on app/code/Magento/InventoryConfiguration/Model/ConfigMinQty.php:
```diff
@@ -0,0 +1,81 @@
+namespace Magento\InventoryConfiguration\Model;
+
+use Magento\Inventory\Model\GetStockItemDataInterface;
+use Magento\InventoryCatalog\Model\GetLegacyStockItem;
```
InventtoryConfigurationApi

Stepa4man replied:
Done

maghamed commented on app/code/Magento/InventoryCatalog/Model/IsBackOrderAvailable.php:
```diff
@@ -0,0 +1,49 @@
+    public function match(string $sku, int $stockItem): bool
+    {
+        $legacyStockItem = $this->getLegacyStockItem->execute($sku);
+        if ($legacyStockItem->getBackorders() != Stock::BACKORDERS_NO) {
```
need to move constant as well

Stepa4man replied:
Done

maghamed commented on app/code/Magento/Inventory/Model/ResourceModel/Stock/GetStockItemData.php:
```diff
@@ -8,17 +8,49 @@
 
 use Magento\Inventory\Model\GetStockItemDataInterface;
+use Magento\Framework\App\ResourceConnection;
+use Magento\InventoryIndexer\Indexer\IndexStructure;
```
break modularity

Stepa4man replied:
Done

Stepa4man committed:
InventoryConfigurationApi

Stepa4man committed:
merge conflicts

Stepa4man committed:
creating interface for getStockItemConfiguration, Removing GetLegacyStockItem Service

Stepa4man committed:
Fixes

maghamed commented on nginx.dev.conf:
```diff
@@ -0,0 +1,204 @@
+## Example configuration:
+# upstream fastcgi_backend {
+#    # use tcp connection
+#    # server  127.0.0.1:9000;
```
no need to commit nginx.dev.conf

Stepa4man committed:
temporary commit with StockItemConf Proxy implementation

Stepa4man committed:
Proxy for StockItemConfiguration

maghamed committed:
Fix tests

aleron75 commented on app/code/Magento/InventoryCatalog/Model/IsBackOrderAvailable.php:
```diff
@@ -0,0 +1,49 @@
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalog\Model;
```
@maghamed: shouldn't this class belong to `InventoryConfiguration` module just like `IsNotManageStock`?

maghamed commented:
#520

maghamed committed:
fix PHP 7.1 syntax

maghamed committed:
fix fatal on getList access

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
Merge remote-tracking branch 'origin/develop' into is_product_salable_conditions_pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

aleron75 committed:
Add new service declaration and remove a plugin 

- Remove plugin definition and class
- Remove plugin's integration test
- Add previous test case in IsProductSalableTest
- Add IsProductSalableForRequestedQtyInterface

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool

Commit:
MSI-565: IsProductSalable conditions pool
